keps/sig-network/2433-topology-aware-hints/README.md

```diff
@@ -0,0 +1,578 @@
+
+```go
+type Endpoint struct {
+  ...
+  // hints contains information associated with how an endpoint should be
+  // consumed.
+  // +optional
+  Hints EndpointHints `json:"hints,omitempty" protobuf:"bytes,7,opt,name=hints"`
+}
+```
+
+```go
+// EndpointHints provides hints describing how an endpoint should be consumed.
+type EndpointHints struct {
+  // zone indicates the zone this endpoint should be consumed by to enable
+  // topology aware routing.
+  // +optional
+  Zone string `json:"ready,omitempty" protobuf:"bytes,1,name=ready"`
+}
+```
+
+#### Future API Expansion
+This approach would allow for future API expansion that enabled specifying
+multiple zones per endpoint with weights. That level of complexity may never be
+necessary, but it will be possible. For example:
+
+```yaml
+topologyHints:
+  zones:
+  - name: example-1a
+    weight: 50
+  - name: example-2a
+    weight: 50
+```
+
+Additionally we could easily expand this API to include support for region
+hints. Although it is unclear if either expansion will be necessary, the API is
+designed in a way to make expansions straightforward.
+
+### Kube-Proxy
+
+When the `TopologyAwareHints` feature gate is enabled, Kube-Proxy will be
+updated to filter endpoints based on topology hints when the following
+conditions are true:
+
+- Kube-Proxy is able to determine the zone it is running within (likely based
+  on node labels).
+- The `topology.kubernetes.io/topology-aware-routing` annotation is set to
+  `Auto` for the Service.
+- At least one endpoint for the Service has a hint pointing to the zone
+  Kube-Proxy is running within.
+- All endpoints for the Service have zone hints.
+
+When the above conditions are true, kube-proxy will only route traffic to
+endpoints with a hint referring to the zone Kube-Proxy is running within.
+
+### EndpointSlice Controller
+
+When the `TopologyAwareHints` feature gate is enabled and the
+`topology.kubernetes.io/topology-aware-routing` annotation is set to `Auto`
+for a Service, the EndpointSlice controller will add hints to EndpointSlices.
+These hints will indicate where an endpoint should be consumed by proxy
+implementations to enable topology aware routing.
+
+The EndpointSlice controller will determine how many endpoints should be
+available for each zone based on the proportion of CPU cores in each zone. If
+it is not possible to determine the number CPU cores, 1 core per node will be
+assumed for calculations.
+
+#### Example
+
+zone-a: 20 CPU cores
+zone-b: 16 CPU cores
+zone-c: 14 CPU cores
+
+In this scenario, the following proportion of endpoints would be allocated for
+each Service:
+
+zone-a: 40%
+zone-b: 32%
+zone-c: 28%
+
+When allocating endpoints to meet this distribution, keeping endpoints in the
+same zone will be prioritized. When same-zone endpoints are exhausted, endpoints
+will be taken from zones that have excess capacity.
+
+#### Overload
+
+Overload is a key concept for this proposal. This occurs when there are less
```

Comment:
this seems to be a property of a zone only, or it will be extended to the other things that you mention like regions?

Reply:
I think the same concepts could apply to region, it's just less common in my experience to see individual clusters spanning zones. This may be something that becomes more relevant with multicluster Services.

Reply:
:+1: